File: Programmin_in_C#-Erasmus/Final_Project/service/UserService.cs
using Final_Project.Domain;
using Final_Project.repository.memoryRepo;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Final_Project.service
{
    public class UserService : BaseService<long, User>
    {
        private Repository<long, Friendship> friendshipRepo;

        public UserService(Repository<long, User> repo, Repository<long, Friendship> frRepo) : base(Validators.ValidatorType.USER, repo)
        {
            friendshipRepo= frRepo;
        }

        /***
         * creates a new instance of the user and adds it to the repo
         * @param email the email of the user
         * @param name the name of the user
         * @throws ValidationException if the user is invalid
         * @throws ExistingException if the entity exists
         */
        public void addUser(string name, string email)
        {
            User user = EntitySingletone.getInstance().createUser(name, email);
            base.addEntity(user);
        }

        /***
         * creates a new instance of the user and adds it to the repo
         * @param ID the id of the user
         * @param email the email of the user
         * @param name the name of the user
         * @throws ValidationException if the user is invalid
         * @throws ExistingException if the entity exists
         */
        public void addUser(long id, string name, string email)
        {
            User user = EntitySingletone.getInstance().createUser(id, name, email);
            base.addEntity(user);
        }

        /***
         * Removes an user and the removes all the friendships related to him
         * @param userId the id of the removed user
         * @return the removed user
         * @throws ExistingException if the entity doesn't exists
         */
        public User deleteUser(long id)
        {
            User user = base.deleteEntity(id);
            friendshipRepo.findAll().ForEach(fr =>
            {
                if(fr.getIdUser1().Equals(id) || fr.getIdUser2().Equals(id))
                {
                    friendshipRepo.delete(fr.getID());
                }
            });
            return user;
        }

        /***
         * updates a user
         * @param ID the id of the user
         * @param email new the email of the user
         * @param name new the name of the user
         * @throws ValidationException if the user is invalid
         * @throws ExistingException if the entity doesn't exists
         */
        public User updateUser(long id, string name, string email)
        {
            User user = EntitySingletone.getInstance().createUser(id, name, email);
            return base.updateEntity(user);
        }
    }
}

File: Programmin_in_Java-Erasmus/Final_Project/src/main/java/server/service/UserService.java
package main.java.server.service;

import main.java.server.domain.Friendship;
import main.java.server.domain.User;
import main.java.server.exceptions.ValidationException;
import main.java.server.exceptions.ExistingException;
import main.java.server.repository.memoryRepo.Repository;
import main.java.server.validators.ValidatorType;

public class UserService extends BaseService<Long, User> {
    private final Repository<Long, Friendship> friendshipRepo;

    public UserService(Repository<Long, User> repository, Repository<Long, Friendship> friendshipRepo) {
        super(ValidatorType.USER, repository);
        this.friendshipRepo = friendshipRepo;
    }

    /***
     * creates a new instance of the user and adds it to the repo
     * @param email the email of the user
     * @param name the name of the user
     * @throws ValidationException if the user is invalid
     * @throws ExistingException if the entity exists
     */

    public Long addUser(String name, String email) {
        User user = EntitySingleton.getInstance().getUser(name, email);
        super.addEntity(user);
        return user.getId();
    }

    /***
     * creates a new instance of the user and adds it to the repo
     * @param ID the id of the user
     * @param email the email of the user
     * @param name the name of the user
     * @throws ValidationException if the user is invalid
     * @throws ExistingException if the entity exists
     */
    public void addUserID(Long ID, String name, String email) {
        //!!!!!!
        //User user = EntitySingleton.getInstance().getUserID(ID, name, email);
        User user = EntitySingleton.getInstance().getUserID(ID, email, name);
        super.addEntity(user);
    }

    /***
     * Removes an user and the removes all the friendships related to him
     * @param userId the id of the removed user
     * @return the removed user
     * @throws ExistingException if the entity doesn't exist
     */
    @Override
    public User deleteEntity(Long userId) {
        User user = super.deleteEntity(userId);
        friendshipRepo.findAll()
                .forEach(friendship -> {
                    if (friendship.getIdUser1().equals(userId) || friendship.getIdUser2().equals(userId)) {
                        friendshipRepo.delete(friendship.getId());
                    }
                });
        return user;
    }

}

